src/Kestrel.Core/Internal/Http2/HPack/HPackDecoder.cs

```diff
@@ -133,6 +139,7 @@ public void OnByte(byte b, IHttpHeadersHandler handler)
                         if (_integerDecoder.BeginDecode((byte)val, IndexedHeaderFieldPrefix))
                         {
                             OnIndexedHeaderField(_integerDecoder.Value, handler);
+                            _headersObserved = true;
```

Comment:
You can consolidate all of this into the Ready state if you move this assignment above this `if` and copy it to the other header types (LiteralHeaderFieldWithIncrementalIndexingRepresentation, etc.). Then it can be enforced once in DynamicTableSizeUpdateRepresentation.

Reply:
Consolidated